src/network/tcpsocket.c

```diff
@@ -81,33 +81,30 @@ static void client_to_server_state(struct ustream *s) {
 
 static void client_read_cb(struct ustream *s, int bytes) {
     char *str;
-    int len;
+    int len = 0;
+    size_t final_len;
 
     do {
         str = ustream_get_read_buf(s, &len);
         if (!str)
             break;
 
         if (network_config.use_symm_enc) {
-            char *base64_dec_str = malloc(B64_DECODE_LEN(strlen(str)));
-            int base64_dec_length = b64_decode(str, base64_dec_str, B64_DECODE_LEN(strlen(str)));
-            char *dec = gcrypt_decrypt_msg(base64_dec_str, base64_dec_length);
-
-            free(base64_dec_str);
+            final_len = *(size_t*)str;
+            if(len < final_len) {//not complete msg, wait for next recv
+                fprintf(stderr,"not complete msg, len:%d, expected len:%d\n", len, final_len);
+                continue;
+            }
+            char *dec = gcrypt_decrypt_msg(str+sizeof(size_t), final_len-sizeof(size_t));
+        
             handle_network_msg(dec);
             free(dec);
+            ustream_consume(s, final_len);//one msg is processed
         } else {
             handle_network_msg(str);
```

Comment:
Why only do this, if symmetric encyption is enabled? Seems like a general problem.

Reply:
fixed